File: backend/functions/src/features/authentication/domain/entities/email_otp_challenge_entity.ts
import {z} from "zod";

/** Desafio de código enviado por e-mail.
 *
 * [codeHash] fica no servidor; o app só envia o código em texto na validação.
 */
export const emailOtpChallengeEntitySchema = z.object({
  id: z.string(),
  userId: z.string(),
  codeHash: z.string(),
  expiresAt: z.coerce.date(),
  attemptCount: z.number().int().nonnegative().default(0),
  consumedAt: z.coerce.date().optional().nullable(),
  createdAt: z.coerce.date(),
});

export type EmailOtpChallengeEntity = z.infer<
  typeof emailOtpChallengeEntitySchema
>;

File: app/lib/features/authentication/domain/entities/email_otp_challenge_entity.dart
import 'package:dart_mappable/dart_mappable.dart';

part 'email_otp_challenge_entity.mapper.dart';

/// Desafio de código enviado por e-mail.
///
/// [codeHash] fica no servidor; o app só envia o código em texto na validação.
@MappableClass()
class EmailOtpChallengeEntity with EmailOtpChallengeEntityMappable {
  EmailOtpChallengeEntity({
    required this.id,
    required this.userId,
    required this.codeHash,
    required this.expiresAt,
    this.attemptCount = 0,
    this.consumedAt,
    required this.createdAt,
  });

  final String id;
  final String userId;
  final String codeHash;
  final DateTime expiresAt;
  final int attemptCount;
  final DateTime? consumedAt;
  final DateTime createdAt;
}
